### PR TITLE
REPL class loader is context class loader

### LINKED ISSUE #14281: Scala 3 REPL does not appear to be loading resources properly
I'm unable to properly load resources from the Scala 3 REPL. Reference project/code is here:
https://github.com/kapunga/whats-up-resources/branches

### core/src/main/scala/wur/Main.scala
```Scala
package wur

object Hello:
  def world: String = io.Source.fromResource("hello.txt").getLines.mkString("\n")

@main def run = println(Hello.world)
```
### core/src/main/resources/hello.txt 
```
Hello,
World!
```

For this project, I can read them just fine when running the project:
```
kapunga ~/whats-up-resources (master)$ sbt core/run
[info] welcome to sbt 1.6.1 (AdoptOpenJDK Java 1.8.0_282)
 ...
[info] running wur.run
Hello,
World!
[success] Total time: 4 s, completed Jan 17, 2022 12:12:32 AM
```

However, when I try and load the resources from the repl, they are not found:

```
kapunga ~/whats-up-resources (master)$ scala3-repl -classpath core/target/scala-3.1.0/core_3-0.1.0-SNAPSHOT.jar
Welcome to Scala 3.1.0 (11.0.10, Java OpenJDK 64-Bit Server VM).
Type in expressions for evaluation. Or try :help.

scala> import wur._

scala> Hello.world
java.io.FileNotFoundException: resource 'hello.txt' was not found in the classpath from the given classloader.
  at scala.io.Source$.fromResource(Source.scala:183)
  at wur.Hello$.world(Main.scala:5)
  ... 34 elided

scala>
```

The resource file is definitely there:
```
kapunga ~/whats-up-resources (master)$ jar tf core/target/scala-3.1.0/core_3-0.1.0-SNAPSHOT.jar
META-INF/MANIFEST.MF
wur/
hello.txt
wur/Hello$.class
wur/Hello.class
wur/Hello.tasty
wur/Main$package$.class
wur/Main$package.class
wur/Main$package.tasty
wur/run.class
wur/run.tasty
``` 

I've confirmed that this is not an issue with SBT or in Scala 2.13.8, where the resource loads just fine.

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes #14281 

Also tweak the test mini-DSL to take context functions.

### CODE REVIEW

**Comment by griggt** [comment]
> Also tweak the test mini-DSL to take context functions.

Thanks!  I had been meaning to PR this.

File: compiler/src/dotty/tools/runner/ScalaClassLoader.scala [hunk 1]
```diff
@@ -14,12 +14,8 @@ import scala.annotation.tailrec
...
 
 final class RichClassLoader(private val self: ClassLoader) extends AnyVal {
-  /** Executing an action with this classloader as context classloader */
-  private def asContext[T](action: => T): T = {
-    val saved = Thread.currentThread.getContextClassLoader
-    try { ScalaClassLoader.setContext(self) ; action }
-    finally ScalaClassLoader.setContext(saved)
-  }
+  /** Execute an action with this classloader as context classloader. */
+  def asContext[T](action: => T): T = ScalaClassLoader.asContext(self)(action)
```
Comment (som-snytt) [inline compiler/src/dotty/tools/runner/ScalaClassLoader.scala:18]: intended to leave that private.

File: compiler/test/dotty/tools/repl/ReplTest.scala [hunk 2]
```diff
@@ -33,11 +34,12 @@ extends ReplDriver(options, new PrintStream(out, true, StandardCharsets.UTF_8.na
   @After def cleanup: Unit =
     storedOutput()
 
-  def fromInitialState[A](op: State => A): A =
-    op(initialState)
+  def initially[A](op: State ?=> A): A = op(using initialState)
+
+  def contextually[A](op: Context ?=> A): A = op(using initialState.context)
```
Comment (som-snytt) [inline compiler/test/dotty/tools/repl/ReplTest.scala:39]: just use `rootCtx`

**Comment by SethTisue** [comment]
@som-snytt is there any possibility this was the way it was on purpose? do you see any downside to changing it to the Scala 2 way?

**Review by anatoliykmetyuk — APPROVED** [review]

**Pull request merged**